gossip/gossip.go

```diff
@@ -0,0 +1,120 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package gossip
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/ava-labs/avalanchego/x/p2p"
+	"github.com/ethereum/go-ethereum/log"
+
+	"github.com/ava-labs/avalanchego/codec"
+	"github.com/ava-labs/avalanchego/ids"
+)
+
+type GossipableAny[T any] interface {
+	*T
+	Gossipable
+}
+
+func NewGossiper[T any, U GossipableAny[T]](
+	set Set[U],
+	client *p2p.Client,
+	codec codec.Manager,
+	codecVersion uint16,
+	gossipSize int,
+	frequency time.Duration,
+) *Gossiper[T, U] {
+	return &Gossiper[T, U]{
+		set:          set,
+		client:       client,
+		codec:        codec,
+		codecVersion: codecVersion,
+		gossipSize:   gossipSize,
+		frequency:    frequency,
+	}
+}
+
+type Gossiper[T any, U GossipableAny[T]] struct {
+	set          Set[U]
+	client       *p2p.Client
+	codec        codec.Manager
+	codecVersion uint16
+	gossipSize   int
+	frequency    time.Duration
+}
+
+func (g *Gossiper[T, U]) Pull(
+	shutdownChan chan struct{},
+	shutdownWg *sync.WaitGroup,
+) {
+	gossipTicker := time.NewTicker(g.frequency)
+	defer func() {
+		gossipTicker.Stop()
+		shutdownWg.Done()
+	}()
+
+	for {
+		select {
+		case <-gossipTicker.C:
+			bloomBytes, err := g.set.GetBloomFilter()
+			if err != nil {
+				log.Warn("failed to marshal bloom filter", "error", err)
+				continue
+			}
+
+			request := PullGossipRequest{
+				BloomFilter: bloomBytes,
+			}
+			msgBytes, err := g.codec.Marshal(g.codecVersion, request)
+			if err != nil {
+				log.Warn("failed to marshal gossip message", "error", err)
+				continue
+			}
+
+			onResponse := func(nodeID ids.NodeID, responseBytes []byte, err error) {
+				if err != nil {
+					log.Warn("failed gossip request", "nodeID", nodeID, "error", err)
+					return
+				}
+
+				response := PullGossipResponse{}
+				if _, err := g.codec.Unmarshal(responseBytes, &response); err != nil {
+					log.Warn("failed to unmarshal gossip", "error", err)
+					return
+				}
+
+				for _, gossipBytes := range response.GossipBytes {
+					gossipable := U(new(T))
```

Comment:
Do you have any reference / documentation on constructing a pointer this way?

Reply:
Would it be simpler to add a `Parser` interface rather than use generic pointer black magic here?

Reply:
What constraints should there be on the size of gossipable bytes and the number of entries in the response if any?

Reply:
> Do you have any reference / documentation on constructing a pointer this way?

First answer on this:

https://stackoverflow.com/questions/69573113/how-can-i-instantiate-a-non-nil-pointer-of-type-argument-with-generic-go

> Would it be simpler to add a Parser interface rather than use generic pointer black magic here?

Yeah I went back and forth on if this is cleaner. It might be because in order to unmarshal/marshal each tx needs a reference to the codec currently. We could probably just have some parse function/interface as well. I'll see what it looks like.